synapse/replication/http/_base.py

```diff
@@ -271,13 +267,11 @@ async def send_request(
                         "Unknown METHOD on %s replication endpoint" % (cls.NAME,)
                     )
 
-                # Here the protocol is hard coded to be http by default or https in case the replication
-                # port is set to have tls true.
-                scheme = "https" if tls else "http"
-                uri = "%s://%s:%s/_synapse/replication/%s/%s" % (
-                    scheme,
-                    host,
-                    port,
+                # Use the instance_map data to retrieve the correct scheme and use the
+                # instance_name to abstract the connection details into the Agent
+                uri = "%s://%s/_synapse/replication/%s/%s" % (
+                    instance_map[instance_name].scheme(),
+                    instance_name,
```

Comment:
Interesting idea. I gave this a go, there is only 1 side-effects AFAICT:
Log lines will show `synapse-replication://` instead of `http://` or `https://`(or `unix://` in the future), so won't know by looking if it's protected by TLS at a glance(it should still be able to be gleaned from other sources in the logs though).

Less of a side-effect and more of an observation:
The possibility of hitting a `SchemeNotSupported` Exception is significantly narrower. All 'scheme's would be programmatically generated, it would error much sooner in the system(at start up, I believe; hinges on if a host:port is missing or not, or a path in future work. Actually, that may have always been the case anyways). In theory, could remove the exception handling for that(but I'm not married to that idea).

Adjusted in [1d1dbff](https://github.com/matrix-org/synapse/pull/15578/commits/1d1dbff9d381efbeecc535cdc214d0412fe96fd4)

Reply:
(Edited main comment to show how the log lines will look now)